src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -335,20 +343,49 @@ public byte[] Http2AltSvcOriginUri
         private ValueTask<(HttpConnectionBase? connection, bool isNewConnection, HttpResponseMessage? failureResponse)>
             GetConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
         {
+            // Do not even attempt at getting/creating a connection if it's already obvious we cannot provided the one requested.
+            if (request.VersionPolicy != HttpVersionPolicy.RequestVersionOrLower)
+            {
+                if (request.Version.Major >= 3 && !_http3Enabled)
+                {
+                    throw new NotSupportedException("ToDo: Message=Requesting HTTP version {0} with version policy {1} while HTTP/{2} is not enabled. See '{3}' AppContext switch.");
+                }
+                if (request.Version.Major >= 2 && !_http2Enabled)
+                {
+                    throw new NotSupportedException("ToDo: Message=Requesting HTTP version {0} with version policy {1} while HTTP/{2} is not enabled. See '{3}' AppContext switch.");
+                }
+            }
+
             if (_http3Enabled && request.Version.Major >= 3)
             {
+                // ToDo: this depends on _assumePrenegotiatedHttp3ForTesting, why is the H3 connection under this condition?
                 HttpAuthority? authority = _http3Authority;
                 if (authority != null)
                 {
                     return GetHttp3ConnectionAsync(request, authority, cancellationToken);
                 }
             }
 
-            if (_http2Enabled && request.Version.Major >= 2)
+            // What about retry? Should we move this check into SendWithRetry?
```

Comment:
I'm not sure what you mean by retry. Can you elaborate?

Reply:
I understand the code better now and got rid of the comment.
But I wasn't sure whether there are any circumstances when we want to retry the request instead of throwing NSE. It was mostly based by not understanding properly [this line](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L787). I checked the code and `RetryOnSameOrNextProxy` is used when we want to try to send with a new connection because we cannot process the request on an existing one and we didn't send anything important yet (`_canRetry` fields).